### PR TITLE
docs: Rook Ceph Upgrade

### PR DESCRIPTION
**Pull request by surajssd**
Add a guide to explain how to upgrade rook-ceph component.

Fixes: https://github.com/kinvolk/lokomotive/issues/906

### CODE REVIEW

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (surajssd) [inline docs/how-to-guides/upgrade-rook-ceph.md:137]: How do we make sure that the CRDs installed here are the version that user is using?

**Comment by surajssd** [comment]
- [ ] Verify with the upstream docs to make sure that we are in sync.

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (invidian) [reply docs/how-to-guides/upgrade-rook-ceph.md:137]: Maybe use `lokoctl component render-manifest rook-ceph | yq 'select( .kind == "CustomResourceDefinition")'`?

BTW, they still didn't migrate to Helm 3... https://github.com/rook/rook/pull/5563

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (invidian) [inline docs/how-to-guides/upgrade-rook-ceph.md:134]: I'd add some brief note here, like "With everything monitored, you can start the update process now." or something.

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (invidian) [inline docs/how-to-guides/upgrade-rook-ceph.md:25]:
```suggestion
Rook Ceph is one of the storage providers of Lokomotive. With a distributed system as complex as
Ceph, the upgrade process might be not trivial. This document enlists steps on how to perform it and
how it should be monitored.
```

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (invidian) [inline docs/how-to-guides/upgrade-rook-ceph.md:29]:
```suggestion
Following steps are inspired by [`rook`](https://rook.io/docs/rook/master/ceph-upgrade.html) docs.
```

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (invidian) [inline docs/how-to-guides/upgrade-rook-ceph.md:31]: First thing first, I'd ensure that cluster is in the healthy state.

**Review by invidian — CHANGES_REQUESTED** [review]
Some nits, otherwise looks nice

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (surajssd) [reply docs/how-to-guides/upgrade-rook-ceph.md]: I have made some changes, PTAL. In the `ceph status` there is an important note which says don't move forward if there is anything other than `HEALTH_OK`.

Comment (invidian) [reply docs/how-to-guides/upgrade-rook-ceph.md]: I meant, before starting Rook Ceph upgrade user should make sure that their Kubernetes cluster is in healthy state.

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (invidian) [inline docs/how-to-guides/upgrade-rook-ceph.md:158]: Broken reference?

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (invidian) [inline docs/how-to-guides/upgrade-rook-ceph.md:167]: Please add some explanation why user should run it. Or what should be done with the output :)

**Review by invidian — CHANGES_REQUESTED** [review]
Looking good, just some nits.

BTW, that's a lot of things to monitor during the upgrade, I wonder if we could simplify that somehow...

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (surajssd) [reply docs/how-to-guides/upgrade-rook-ceph.md:167]: This was more of a generic verify things are ok kinda commands. Without giving much explanation into what to expect. Because the output might change dependent on the cluster.

Let me produce some intelligent sentences around it!

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (invidian) [inline docs/how-to-guides/upgrade-rook-ceph.md:25]: I'd link here to the component documentation.

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (invidian) [inline docs/how-to-guides/upgrade-rook-ceph.md:35]:
```suggestion
Start a shell in the toolbox pod as specified in [this
```

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (invidian) [inline docs/how-to-guides/upgrade-rook-ceph.md:37]:
```suggestion
doc](rook-ceph-storage.md#enable-and-access-toolbox) and run the
following command:
```

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (invidian) [inline docs/how-to-guides/upgrade-rook-ceph.md:33]:
```suggestion
### Step 1: Ensure that all pools have the `AUTOSCALE` plugin enabled
```

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (invidian) [inline docs/how-to-guides/upgrade-rook-ceph.md:47]: Actually, why this is important from the update process point of view? Will data in the cluster increase because of the update? If so, how much etc?

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (invidian) [inline docs/how-to-guides/upgrade-rook-ceph.md:46]:
```suggestion
column says `warn`, then run the command below, to make sure that pool autoscaling is enabled.
```

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (invidian) [inline docs/how-to-guides/upgrade-rook-ceph.md:50]: This should be executed for all pools I suppose.

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (invidian) [inline docs/how-to-guides/upgrade-rook-ceph.md:65]: Perhaps we could replace `looks fine` here with something less ambiguous.

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (invidian) [inline docs/how-to-guides/upgrade-rook-ceph.md:62]: I suppose output of this command will be changing during the upgrade. I think we should mention that. And that it should go back to `HEALTH_OK` when the upgrade is finished.

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (invidian) [inline docs/how-to-guides/upgrade-rook-ceph.md:74]: Why? "So you can see if the update has been triggered and see it's progress" ? Or "when update process is started, you should start seeing pods being re-created etc."

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (invidian) [inline docs/how-to-guides/upgrade-rook-ceph.md:95]: Should we mention specific version here?

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (invidian) [inline docs/how-to-guides/upgrade-rook-ceph.md:107]: Should we mention specific version here?

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (invidian) [inline docs/how-to-guides/upgrade-rook-ceph.md:111]: As a person who does the upgrade, what kind of events are expected here? Or what events should you specifically watch out for?

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (invidian) [inline docs/how-to-guides/upgrade-rook-ceph.md:141]: Why this is important?

**Review by invidian — CHANGES_REQUESTED** [review]
Looking good. BTW, comment about the CRDs is not addressed still.

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (surajssd) [reply docs/how-to-guides/upgrade-rook-ceph.md:47]: The data won't increase but if there is an imbalance in number of OSDs for whatever reason and when the data starts to replicate that is missing from the unresponsive OSD then we need to make sure that PGs can autoscale otherwise the data won't be replicated. This is a situation specially when the cluster is full over 60%.

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (surajssd) [reply docs/how-to-guides/upgrade-rook-ceph.md:33]: It can be enabled without being `on` so I have rephrased it.

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (surajssd) [reply docs/how-to-guides/upgrade-rook-ceph.md:50]: We currently have only one pool created with the name `replicapool` so only above command should suffice.

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (invidian) [reply docs/how-to-guides/upgrade-rook-ceph.md:47]: What's the OSD time to trigger data replication to new OSD? As far as I know, there is some threshold for transient OSD failures to avoid stressing the cluster with recovery process.

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (surajssd) [reply docs/how-to-guides/upgrade-rook-ceph.md:74]: So that none of the pods go into `CrashLoopBackOff`. Yes mostly to monitor they start correctly.

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (surajssd) [reply docs/how-to-guides/upgrade-rook-ceph.md:95]: Yep, this is an upgrade doc. So anyone using latest release of lokoctl should expect above version.

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (surajssd) [reply docs/how-to-guides/upgrade-rook-ceph.md:111]: No nothing specific, to watch here. But in general issues like ImagePullBackOffs, PSP issues, CrashLoopBackOffs are easily tracked in events.

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (surajssd) [reply docs/how-to-guides/upgrade-rook-ceph.md:141]: To verify that the images are updated successfully at the end of the upgrade process.

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (invidian) [reply docs/how-to-guides/upgrade-rook-ceph.md:95]: I thought we make document version-independent. This means whenever we update Rook Ceph, we should update this document as well? This perhaps also apply to #1163 then :thinking:

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (invidian) [reply docs/how-to-guides/upgrade-rook-ceph.md:144]: We should write that down then :)

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (invidian) [reply docs/how-to-guides/upgrade-rook-ceph.md:114]: We should write that down then :)

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (invidian) [reply docs/how-to-guides/upgrade-rook-ceph.md:77]: We should write that down :)

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (surajssd) [reply docs/how-to-guides/upgrade-rook-ceph.md:156]: Ideally the solution to this problem is that lokoctl spits out the CRD config or applies them manually before any component upgrade.

But right now we may rely on `yq`, although I don't like adding another dependency.

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (surajssd) [reply docs/how-to-guides/upgrade-rook-ceph.md:98]: Yep this applies to both the docs.

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (surajssd) [reply docs/how-to-guides/upgrade-rook-ceph.md]: In a cluster of three node worker pool with replica value set to three, during the upgrade you may notice that when even one OSD goes down (for upgrade), the data replication is triggered immediately. You can see IO spike in the `ceph status` output.

Comment (invidian) [reply docs/how-to-guides/upgrade-rook-ceph.md]: Interesting. In any case, we should write this down :)

Comment (surajssd) [reply docs/how-to-guides/upgrade-rook-ceph.md]: Yes it is written.

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (invidian) [inline docs/how-to-guides/upgrade-rook-ceph.md:21]: Header must be updated to new docs format. See other howtos as example.

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (invidian) [reply docs/how-to-guides/upgrade-rook-ceph.md:160]: So maybe we should change it to v0.5.0 for now?

Hmm, we could perhaps add a flag or something to update the CRDs :thinking:

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (invidian) [inline docs/how-to-guides/upgrade-rook-ceph.md:25]: In OpenEBS guide we recommend scheduling a downtime and taking backup before upgrading storage. Should we do the same here?

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (invidian) [inline docs/how-to-guides/upgrade-rook-ceph.md:46]:
```suggestion
required to ensure that the placement groups scale up as the data in the cluster increases during the upgrade when individual OSDs gets updated.
```

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (invidian) [inline docs/how-to-guides/upgrade-rook-ceph.md:71]:
```suggestion
state. If the cluster is more than 60% full, then the output can sometimes turn into `HEALTH_WARN`.
```

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (invidian) [inline docs/how-to-guides/upgrade-rook-ceph.md:86]: This sentence doesn't sound right to me.
```suggestion
Run the following command in a new terminal window to keep an eye on the rook version as it
is updated for all the sub-components:
```

**Review by invidian — CHANGES_REQUESTED** [review]
Looking good.

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (surajssd) [reply docs/how-to-guides/upgrade-rook-ceph.md:46]: This should always be `on` regardless of upgrade or anything else.

Comment (invidian) [reply docs/how-to-guides/upgrade-rook-ceph.md:38]: This part still feels somehow incomplete to me. Maybe we can add that "it is recommended that this option is always on, especially during the upgrade when..." ?

**Review by invidian — DISMISSED** [review]
Just one last nit, but I think we could merge regardless. Nice work @surajssd

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (iaguis) [inline docs/how-to-guides/upgrade-rook-ceph.md:23]:
```suggestion
The following steps are inspired by the [`rook`](https://rook.io/docs/rook/master/ceph-upgrade.html) docs.
```

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (iaguis) [inline docs/how-to-guides/upgrade-rook-ceph.md:37]:
```suggestion
column says `warn`, then run the command below to make sure that pool autoscaling is enabled. It is
```

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (iaguis) [inline docs/how-to-guides/upgrade-rook-ceph.md:63]: So what happens if during the upgrade it shows `HEALTH_WARN` is it ok if it doesn't get full or does everything explode after that? What advice can we give to the user?

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (iaguis) [inline docs/how-to-guides/upgrade-rook-ceph.md:68]: I find this phrasing weird.

```suggestion
Open another terminal window and keep an eye on the `STATUS` field of the following output. Make sure that the pods are restarted in record time and don't go into
```

Comment (iaguis) [reply docs/how-to-guides/upgrade-rook-ceph.md:68]: Also, what does "pods are restated in record time" mean?

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (iaguis) [inline docs/how-to-guides/upgrade-rook-ceph.md:123]:
```suggestion
> **NOTE**: Accessing the dashboard can be a hassle because while the components are upgrading you may
```

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (iaguis) [inline docs/how-to-guides/upgrade-rook-ceph.md:152]: Is the idea to keep this tag updated as we release?

**Review by iaguis — CHANGES_REQUESTED** [review]
Some comments. One general concern is that we keep telling the user to monitor updates but we don't say what to do if something goes wrong. I don't know if there's much we can say about that but it feels weird.

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (surajssd) [reply docs/how-to-guides/upgrade-rook-ceph.md:68]: The time after which the controller times out (this can be for various reasons) / starts showing error / restarts pods for some failure condition, etc.

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (surajssd) [reply docs/how-to-guides/upgrade-rook-ceph.md:152]: We have to update the doc on each upgrade of Rook Ceph. Things may change between rook-ceph releases.

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (surajssd) [reply docs/how-to-guides/upgrade-rook-ceph.md:63]: HEALTH_WARN is ok to be in. TBH the user does not have control over the upgrade process and if the upgrade fails for whatever reason we can have look into it case by case basis. This upgrade doc should reflect such cases if there are any coming from upstream.

**Comment by surajssd** [comment]
> One general concern is that we keep telling the user to monitor updates but we don't say what to do if something goes wrong. I don't know if there's much we can say about that but it feels weird.

I think all this visibility helps user to connect dots in an automated upgrade process, if something were to go wrong. A user can simply run `lokoctl component apply rook rook-ceph` and might wonder what is going on under or why cluster won't store data all of sudden.

Expecting failures will be hard, but if there are known issues we should always document it in this document. But we provide the user doing upgrade enough visibility and tools to dig deeper if something were to go wrong.

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (surajssd) [reply docs/how-to-guides/upgrade-rook-ceph.md:38]: Could you quote what should I change it to exactly? I am finding it hard to rephrase.

**Review by invidian — APPROVED** [review]
LGTM, but we should perhaps wait for @iaguis approval.

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (iaguis) [inline docs/how-to-guides/upgrade-rook-ceph.md:39]: I think something like this should do the trick.

```suggestion
Ensure that the `AUTOSCALE` column outputs `on` and not `warn`. This should always be `on` but it's especially important for upgrades. If the output of the `AUTOSCALE`
column says `warn`, then run the command below to make sure that pool autoscaling is enabled. It is
required to ensure that the placement groups scale up as the data in the cluster increases.
```

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (iaguis) [reply docs/how-to-guides/upgrade-rook-ceph.md:64]: Then I guess we shouldn't say "make sure that the output stays in `HEALTH_OK`" because it's not actionable for the user. I'll make a suggestion.

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (iaguis) [inline docs/how-to-guides/upgrade-rook-ceph.md:64]: How about this?

```suggestion
During the ongoing upgrade and after completion, the output should stay in `HEALTH_OK`
state, although if the cluster is more than 60% full, it can sometimes turn into `HEALTH_WARN` temporarily.
```

File: docs/how-to-guides/upgrade-rook-ceph.md
Comment (iaguis) [inline docs/how-to-guides/upgrade-rook-ceph.md:69]: I still don't know what `in record time` means. To me it means "in the least amount of time ever" which doesn't make sense in this context. Maybe something like this?

```suggestion
sure that the pods are restarted correctly and don't go into `CrashLoopBackOff` state. Leave
```

**Review by iaguis — CHANGES_REQUESTED** [review]
Some small changes.

**Review by iaguis — APPROVED** [review]
lgtm

**Pull request merged**
